specs/Reservations-specs.rb

```diff
@@ -0,0 +1,232 @@
+  end
+
+  describe 'all reservations' do
+    it "has a collection of reservations" do
+      @new_hotel.all_reservations.must_be_kind_of Array
+    end
+  end
+
+  describe 'new reservation' do
+    before do
+      @new_hotel.clear_reservations
+      @new_hotel1 = Hotel::Reservations.new
+    end
+    it 'must create a new booking' do
+      new_booking1 = @new_hotel1.new_reservation("2017-09-21", "2017-09-23", 1)
+      new_booking1.must_be_instance_of Hotel::Booking
+    end
+    it 'must raise an error if the check-in date is in the past' do
+
+      proc{new_booking = @new_hotel1.new_reservation("2017-06-01", "2017-09-21", 2)}.must_raise Hotel::InvalidDateError
+    end
+    it 'must raise an error if the check-in date is the same as the check-out date' do
+      proc{new_booking = @new_hotel1.new_reservation("2017-09-20", "2017-09-20"), 3}.must_raise Hotel::InvalidDateError
+    end
+    it 'must raise an error if the check_in date is after the check-out date' do
+      proc{new_booking = @new_hotel1.new_reservation("2017-09-20", "2017-09-18"), 4}.must_raise Hotel::InvalidDateError
+    end
+    it 'must raise an error if the date is not valid on calendar' do
+      proc{new_booking = @new_hotel1.new_reservation("2018-02-30", "2018-02-31"), 5}.must_raise ArgumentError
+    end
+    it 'allows a new reservation to be made on a room on the same day as previous check-out' do
+      @new_hotel1.clear_reservations
+      @new_reservation1 = @new_hotel1.new_reservation("2018-01-01", "2018-01-05", 1)
+      @new_reservation2 = @new_hotel1.new_reservation("2018-01-05", "2018-01-07", 1)
+      @new_hotel1.all_reservations.length.must_equal 2
+    end
+  end
+
+  describe 'check availability' do
+    before do
+      @new_hotel.clear_reservations
+      @new_reservation1 = @new_hotel.new_reservation("2018-01-01", "2018-01-05", 1)
+    end
+    # after do
+    #   @new_hotel.clear_reservations
+    # end
+    it 'must must raise argument error if room is not available' do
+      proc{ new_reservation2 = @new_hotel.new_reservation("2018-01-01", "2018-01-04", 1) }.must_raise ArgumentError
+    end
+  end
+
+  describe 'list rooms available by date' do
+    before do
+      @new_hotel.clear_reservations
+      @new_reservation1 = @new_hotel.new_reservation("2018-01-01", "2018-01-05", 1)
+      @new_reservation2 = @new_hotel.new_reservation("2018-01-01", "2018-01-05", 2)
+      @new_reservation3 = @new_hotel.new_reservation("2018-01-01", "2018-01-05", 10)
+    end
+    after do
+      @new_hotel.clear_reservations
+    end
+    it 'must return an array' do
+      @new_hotel.list_rooms_available_by_date("2018-01-02").must_be_kind_of Array
+    end
+    it 'must return the correct number of rooms available' do
+      @new_hotel.list_rooms_available_by_date("2018-01-03").length.must_equal 17
+    end
+    it 'must return the correct rooms available' do
+      @new_hotel.list_rooms_available_by_date("2018-01-03")[0].room_number.must_equal 3
+    end
```

Comment:
Other cases I'd be interested in:
- What if there are no reservations? (it should return a list of all the rooms)
- Do reservations that don't overlap the date range affect the list? (they shouldn't)
  - Check the boundaries of what counts as an overlap. What about a reservation that ends on the checkin date, or starts on the checkout date?
- What if there are no vacancies? (it should return an empty array)
- What if your date range spans several small reservations? (all of the rooms from those reservations should be omitted from the list)